cmd/nephe-controller/secret_webhook.go

```diff
@@ -0,0 +1,169 @@
+	isDelete bool) (error, *crdv1alpha1.CloudProviderAccount) {
+	// TODO: Figure out how to get all CRs
+	cpaList := &crdv1alpha1.CloudProviderAccountList{}
+	options := &controllerclient.ListOptions{Limit: 1000}
+	err := v.Client.List(context.TODO(), cpaList, options)
+	if err != nil {
+		return fmt.Errorf("failed to get CloudProviderAccount list, err:%v", err), nil
+	}
+
+	for _, cpa := range cpaList.Items {
+		v.Log.Info("Checking CloudProvideAccount", "Name", cpa.Name, "Namespace", cpa.Namespace)
+		if cpa.Spec.AWSConfig != nil {
+			if cpa.Spec.AWSConfig.SecretRef.Name == oldSecret.Name && cpa.Spec.AWSConfig.SecretRef.Namespace == oldSecret.Namespace {
+				if isDelete {
+					// Delete is not allowed.
+					v.Log.Info("Cannot delete", "Secret", oldSecret.Name, "CloudProvideAccount", cpa.Name)
+					return fmt.Errorf("cannot delete Secret %v", oldSecret.Name), &cpa
+				}
+
+				if ok := v.allowSecretUpdate(newSecret, oldSecret, cpa.Spec.AWSConfig.SecretRef.Key); !ok {
+					return fmt.Errorf("data %v cannot be changed for Secret %v",
+						cpa.Spec.AWSConfig.SecretRef.Key, oldSecret.Name), &cpa
+				}
+			}
+		}
+
+		if cpa.Spec.AzureConfig != nil {
+			if cpa.Spec.AzureConfig.SecretRef.Name == oldSecret.Name && cpa.Spec.AzureConfig.SecretRef.Namespace == oldSecret.Namespace {
+				if isDelete {
+					// Delete is not allowed.
+					v.Log.Info("Cannot delete", "Secret", oldSecret.Name, "CloudProviderAccount", oldSecret.Name)
+					return fmt.Errorf("cannot delete Secret %v", oldSecret.Name), &cpa
+				}
+				if ok := v.allowSecretUpdate(newSecret, oldSecret, cpa.Spec.AzureConfig.SecretRef.Key); !ok {
+					return fmt.Errorf("data %v cannot be changed for Secret %v",
+						cpa.Spec.AzureConfig.SecretRef.Key, oldSecret.Name), &cpa
+				}
+			}
+		}
+	}
+	return nil, nil
+}
+
+func (v *SecretValidator) validateCreate(req admission.Request) admission.Response {
+	newSecret := &corev1.Secret{}
+	err := v.decoder.Decode(req, newSecret)
+	if err != nil {
+		v.Log.Error(err, "Failed to decode Secret", "SecretValidator", req.Name)
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	return admission.Allowed("")
+}
+
+func (v *SecretValidator) validateUpdate(req admission.Request) admission.Response {
+	newSecret := &corev1.Secret{}
+	err := v.decoder.Decode(req, newSecret)
+	if err != nil {
+		v.Log.Error(err, "Failed to decode Secret", "SecretValidator", req.Name)
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	oldSecret := &corev1.Secret{}
+	if req.OldObject.Raw != nil {
+		if err := json.Unmarshal(req.OldObject.Raw, &oldSecret); err != nil {
+			v.Log.Error(err, "Failed to decode old Secret", "SecretValidator", req.Name)
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+	}
+
+	err, cpa := v.getCPABySecret(newSecret, oldSecret, false)
+	if err != nil && strings.Contains(err.Error(), "failed to get CloudProviderAccount list") {
```

Comment:
you are doing pattern match on error. may be define a string const and use at both places..

Reply:
Done